### ISSUE TITLE
Robustly clean up and restore state after tests

### ISSUE THREAD

**Issue by nfischer** [shelljs/shelljs#834]
Inspired by #833. We should create a robust solution to ensure each test runs with fresh state, and we successfully cleanup after each test.

Right now, some tests can `cd()` to change the working directory, and we won't properly cleanup after them. As an example, here's a bad `afterEach()` which assumes the test doesn't change working directory:

https://github.com/shelljs/shelljs/blob/9035b27403c7e3c3dc801939d20dec8fe6143c06/test/ln.js#L18-L20

Another issue is that `t.context.tmp` usually refers to a **relative path** returned by `utils.getTempDir()`. I think we can be more robust by returning an absolute path.

---

We should create utility functions to do robust setup and teardown, to be used by every test suite (and the test suites can extend this to provide additional setup).

@freitagbr any additional thoughts?

**Comment by freitagbr** [comment]
I agree that `t.context.tmp` should contain the full path of the temporary directory, not the relative. But, the setup and teardown functions should take care of setting all of this up. The `setup` and `teardown` functions might look like the `beforeEach` and `afterEach` in `test/ln.js`, for example. Then, in every test, we could just put `test.beforeEach(setup)` and `test.afterEach(teardown)`. I don't know if ava has a way to do this globally for every test.

Another improvement might be using a dedicated temp directory module like [`tmp`](https://www.npmjs.com/package/tmp) or [`tempy`](https://www.npmjs.com/package/tempy), though our solution should continue to work fine.

**Comment by nfischer** [comment]
> I don't know if ava has a way to do this globally for every test.

@freitagbr Huh? `beforeEach()` does run before every test, as-is. What did you mean?

**Comment by nfischer** [comment]
I'll write up a PR in the next couple days.

**Comment by freitagbr** [comment]
> Huh? beforeEach() does run before every test, as-is.

`beforeEach()` runs before every test in that file, but not for every file. Some test frameworks have a way to specify a `beforeEach` for every test in every file, but from reading the documentation, I don't think ava has a way to do this. But that's fine, it's not a big deal if each test file has `beforeEach` in it.

**Comment by nfischer** [comment]
Another thing (caught this in #874), we should robustly set `process.env` to its original value (resetting each property). This should happen in `afterEach.always`.
